File: src/main/java/com/smarthome/backend/server/api/services/modules/HueModuleService.java
package com.smarthome.backend.server.api.services.modules;

import java.io.IOException;
import java.util.Map;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.google.gson.Gson;
import com.smarthome.backend.server.actions.ActionManager;
import com.smarthome.backend.server.api.ApiRouter;
import com.smarthome.backend.server.api.modules.hue.HueModuleManager;
import com.smarthome.backend.server.db.DatabaseManager;
import com.smarthome.backend.server.events.EventStreamManager;

/**
 * Service für Hue-Module-API-Endpunkte.
 */
public class HueModuleService {
    private static final Logger logger = LoggerFactory.getLogger(HueModuleService.class);
    private static final Gson gson = new Gson();
    
    private final HueModuleManager hueModule;
    
    public HueModuleService(DatabaseManager databaseManager, EventStreamManager eventStreamManager, ActionManager actionManager ) {
        this.hueModule = new HueModuleManager(databaseManager, eventStreamManager, actionManager);
    }
    
    public void handleRequest(com.sun.net.httpserver.HttpExchange exchange, String method, String path) throws IOException {
        logger.debug("HueModuleService: {} {}", method, path);
        
        if (path.equals("/modules/hue/bridges/discover")) {
            if ("GET".equals(method)) {
                discoverBridges(exchange);
            } else {
                ApiRouter.sendResponse(exchange, 405, gson.toJson(Map.of("error", "Method not allowed")));
            }
        } else if (path.matches("/modules/hue/bridges/[^/]+/pair")) {
            if ("POST".equals(method)) {
                String bridgeId = ApiRouter.extractPathParam(path, "/modules/hue/bridges/{bridgeId}/pair");
                pairBridge(exchange, bridgeId);
            } else {
                ApiRouter.sendResponse(exchange, 405, gson.toJson(Map.of("error", "Method not allowed")));
            }
        } else if (path.matches("/modules/hue/discover/devices/[^/]+")) {
            if ("GET".equals(method)) {
                String bridgeId = ApiRouter.extractPathParam(path, "/modules/hue/discover/devices/{bridgeId}");
                discoverDevices(exchange, bridgeId);
            } else {
                ApiRouter.sendResponse(exchange, 405, gson.toJson(Map.of("error", "Method not allowed")));
            }
        } else if (path.matches("/modules/hue/devices/[^/]+/setSensitivity")) {
            if ("PUT".equals(method) || "POST".equals(method)) {
                String deviceId = ApiRouter.extractPathParam(path, "/modules/hue/devices/{deviceId}/setSensitivity");
                setSensitivity(exchange, deviceId);
            } else {
                ApiRouter.sendResponse(exchange, 405, gson.toJson(Map.of("error", "Method not allowed")));
            }
        } else if (path.matches("/modules/hue/devices/[^/]+/setOn")) {
            if ("PUT".equals(method) || "POST".equals(method)) {
                String deviceId = ApiRouter.extractPathParam(path, "/modules/hue/devices/{deviceId}/setOn");
                setOn(exchange, deviceId);
            } else {
                ApiRouter.sendResponse(exchange, 405, gson.toJson(Map.of("error", "Method not allowed")));
            }
        } else if (path.matches("/modules/hue/devices/[^/]+/setOff")) {
            if ("PUT".equals(method) || "POST".equals(method)) {
                String deviceId = ApiRouter.extractPathParam(path, "/modules/hue/devices/{deviceId}/setOff");
                setOff(exchange, deviceId);
            } else {
                ApiRouter.sendResponse(exchange, 405, gson.toJson(Map.of("error", "Method not allowed")));
            }
        } else if (path.matches("/modules/hue/devices/[^/]+/setBrightness")) {
            if ("PUT".equals(method) || "POST".equals(method)) {
                String deviceId = ApiRouter.extractPathParam(path, "/modules/hue/devices/{deviceId}/setBrightness");
                setBrightness(exchange, deviceId);
            } else {
                ApiRouter.sendResponse(exchange, 405, gson.toJson(Map.of("error", "Method not allowed")));
            }
        } else if (path.matches("/modules/hue/devices/[^/]+/setTemperature")) {
            if ("PUT".equals(method) || "POST".equals(method)) {
                String deviceId = ApiRouter.extractPathParam(path, "/modules/hue/devices/{deviceId}/setTemperature");
                setTemperature(exchange, deviceId);
            } else {
                ApiRouter.sendResponse(exchange, 405, gson.toJson(Map.of("error", "Method not allowed")));
            }
        } else if (path.matches("/modules/hue/devices/[^/]+/setColor")) {
            if ("PUT".equals(method) || "POST".equals(method)) {
                String deviceId = ApiRouter.extractPathParam(path, "/modules/hue/devices/{deviceId}/setColor");
                setColor(exchange, deviceId);
            } else {
                ApiRouter.sendResponse(exchange, 405, gson.toJson(Map.of("error", "Method not allowed")));
            }
        } else {
            ApiRouter.sendResponse(exchange, 404, gson.toJson(Map.of("error", "Endpoint not found")));
        }
    }
    
    private void discoverBridges(com.sun.net.httpserver.HttpExchange exchange) throws IOException {
        hueModule.discoverBridges(exchange);
    }
    
    /**
     * API-Endpunkt zum Pairing mit einer Hue Bridge.
     * Die bridgeId wird aus dem Pfad extrahiert.
     * Erwartet im Body optional: deviceType und generateClientKey.
     */
    private void pairBridge(com.sun.net.httpserver.HttpExchange exchange, String bridgeId) throws IOException {
        hueModule.pairBridge(exchange, bridgeId);
    }
    
    /**
     * API-Endpunkt zum Suchen von Geräten einer Hue Bridge.
     * Die bridgeId wird aus dem Pfad extrahiert.
     */
    private void discoverDevices(com.sun.net.httpserver.HttpExchange exchange, String bridgeId) throws IOException {
        hueModule.discoverDevices(exchange, bridgeId);
    }
    
    /**
     * API-Endpunkt zum Setzen der Sensitivity eines Hue Motion Sensors.
     * Die deviceId wird aus dem Pfad extrahiert.
     * Erwartet im Body: {"sensitivity": <int>}
     */
    private void setSensitivity(com.sun.net.httpserver.HttpExchange exchange, String deviceId) throws IOException {
        logger.info("Setze Sensitivity für Hue Motion Sensor: {}", deviceId);
        String requestBody = ApiRouter.readRequestBody(exchange);
        
        try {
            @SuppressWarnings("unchecked")
            Map<String, Object> request = gson.fromJson(requestBody, Map.class);
            
            // Prüfe, ob sensitivity im Request vorhanden ist
            if (!request.containsKey("sensitivity")) {
                ApiRouter.sendResponse(exchange, 400, gson.toJson(Map.of("error", "sensitivity parameter is required")));
                return;
            }
            
            Object sensitivityObj = request.get("sensitivity");
            int sensitivity;
            if (sensitivityObj instanceof Number) {
                sensitivity = ((Number) sensitivityObj).intValue();
            } else if (sensitivityObj instanceof String) {
                sensitivity = Integer.parseInt((String) sensitivityObj);
            } else {
                ApiRouter.sendResponse(exchange, 400, gson.toJson(Map.of("error", "sensitivity must be a number")));
                return;
            }
            
            // Rufe die Methode im HueModuleManager auf
            boolean success = hueModule.setSensitivity(deviceId, sensitivity);
            if (success) {
                ApiRouter.sendResponse(exchange, 200, gson.toJson(Map.of("success", true)));
            } else {
                ApiRouter.sendResponse(exchange, 404, gson.toJson(Map.of("error", "Gerät nicht gefunden")));
            }
        } catch (Exception e) {
            logger.error("Fehler beim Setzen der Sensitivity", e);
            ApiRouter.sendResponse(exchange, 400, gson.toJson(Map.of("error", "Invalid request: " + e.getMessage())));
        }
    }
    
    /**
     * API-Endpunkt zum Einschalten eines Hue Light.
     * Die deviceId wird aus dem Pfad extrahiert.
     * Kein Body erforderlich.
     */
    private void setOn(com.sun.net.httpserver.HttpExchange exchange, String deviceId) throws IOException {
        logger.info("Schalte Hue Light ein: {}", deviceId);
        
        try {
            // Rufe die Methode im HueModuleManager auf
            boolean success = hueModule.setOn(deviceId);
            if (success) {
                ApiRouter.sendResponse(exchange, 200, gson.toJson(Map.of("success", true)));
            } else {
                ApiRouter.sendResponse(exchange, 404, gson.toJson(Map.of("error", "Gerät nicht gefunden oder nicht unterstützt")));
            }
        } catch (Exception e) {
            logger.error("Fehler beim Einschalten des Hue Light", e);
            ApiRouter.sendResponse(exchange, 400, gson.toJson(Map.of("error", "Invalid request: " + e.getMessage())));
        }
    }
    
    /**
     * API-Endpunkt zum Ausschalten eines Hue Light.
     * Die deviceId wird aus dem Pfad extrahiert.
     * Kein Body erforderlich.
     */
    private void setOff(com.sun.net.httpserver.HttpExchange exchange, String deviceId) throws IOException {
        logger.info("Schalte Hue Light aus: {}", deviceId);
        
        try {
            // Rufe die Methode im HueModuleManager auf
            boolean success = hueModule.setOff(deviceId);
            if (success) {
                ApiRouter.sendResponse(exchange, 200, gson.toJson(Map.of("success", true)));
            } else {
                ApiRouter.sendResponse(exchange, 404, gson.toJson(Map.of("error", "Gerät nicht gefunden oder nicht unterstützt")));
            }
        } catch (Exception e) {
            logger.error("Fehler beim Ausschalten des Hue Light", e);
            ApiRouter.sendResponse(exchange, 400, gson.toJson(Map.of("error", "Invalid request: " + e.getMessage())));
        }
    }
    
    /**
     * API-Endpunkt zum Setzen der Helligkeit eines Hue Light.
     * Die deviceId wird aus dem Pfad extrahiert.
     * Erwartet im Body: {"brightness": <int>}
     */
    private void setBrightness(com.sun.net.httpserver.HttpExchange exchange, String deviceId) throws IOException {
        logger.info("Setze Helligkeit für Hue Light: {}", deviceId);
        String requestBody = ApiRouter.readRequestBody(exchange);
        
        try {
            @SuppressWarnings("unchecked")
            Map<String, Object> request = gson.fromJson(requestBody, Map.class);
            
            // Prüfe, ob brightness im Request vorhanden ist
            if (!request.containsKey("brightness")) {
                ApiRouter.sendResponse(exchange, 400, gson.toJson(Map.of("error", "brightness parameter is required")));
                return;
            }
            
            Object brightnessObj = request.get("brightness");
            int brightness;
            if (brightnessObj instanceof Number) {
                brightness = ((Number) brightnessObj).intValue();
            } else if (brightnessObj instanceof String) {
                brightness = Integer.parseInt((String) brightnessObj);
            } else {
                ApiRouter.sendResponse(exchange, 400, gson.toJson(Map.of("error", "brightness must be a number")));
                return;
            }
            
            // Rufe die Methode im HueModuleManager auf
            boolean success = hueModule.setBrightness(deviceId, brightness);
            if (success) {
                ApiRouter.sendResponse(exchange, 200, gson.toJson(Map.of("success", true)));
            } else {
                ApiRouter.sendResponse(exchange, 404, gson.toJson(Map.of("error", "Gerät nicht gefunden oder nicht unterstützt")));
            }
        } catch (Exception e) {
            logger.error("Fehler beim Setzen der Helligkeit", e);
            ApiRouter.sendResponse(exchange, 400, gson.toJson(Map.of("error", "Invalid request: " + e.getMessage())));
        }
    }
    
    /**
     * API-Endpunkt zum Setzen der Farbtemperatur eines Hue Light.
     * Die deviceId wird aus dem Pfad extrahiert.
     * Erwartet im Body: {"temperature": <int>}
     */
    private void setTemperature(com.sun.net.httpserver.HttpExchange exchange, String deviceId) throws IOException {
        logger.info("Setze Farbtemperatur für Hue Light: {}", deviceId);
        String requestBody = ApiRouter.readRequestBody(exchange);
        
        try {
            @SuppressWarnings("unchecked")
            Map<String, Object> request = gson.fromJson(requestBody, Map.class);
            
            // Prüfe, ob temperature im Request vorhanden ist
            if (!request.containsKey("temperature")) {
                ApiRouter.sendResponse(exchange, 400, gson.toJson(Map.of("error", "temperature parameter is required")));
                return;
            }
            
            Object temperatureObj = request.get("temperature");
            int temperature;
            if (temperatureObj instanceof Number) {
                temperature = ((Number) temperatureObj).intValue();
            } else if (temperatureObj instanceof String) {
                temperature = Integer.parseInt((String) temperatureObj);
            } else {
                ApiRouter.sendResponse(exchange, 400, gson.toJson(Map.of("error", "temperature must be a number")));
                return;
            }
            
            // Rufe die Methode im HueModuleManager auf
            boolean success = hueModule.setTemperature(deviceId, temperature);
            if (success) {
                ApiRouter.sendResponse(exchange, 200, gson.toJson(Map.of("success", true)));
            } else {
                ApiRouter.sendResponse(exchange, 404, gson.toJson(Map.of("error", "Gerät nicht gefunden oder nicht unterstützt")));
            }
        } catch (Exception e) {
            logger.error("Fehler beim Setzen der Farbtemperatur", e);
            ApiRouter.sendResponse(exchange, 400, gson.toJson(Map.of("error", "Invalid request: " + e.getMessage())));
        }
    }
    
    /**
     * API-Endpunkt zum Setzen der Farbe eines Hue Light.
     * Die deviceId wird aus dem Pfad extrahiert.
     * Erwartet im Body: {"x": <int>, "y": <int>}
     */
    private void setColor(com.sun.net.httpserver.HttpExchange exchange, String deviceId) throws IOException {
        logger.info("Setze Farbe für Hue Light: {}", deviceId);
        String requestBody = ApiRouter.readRequestBody(exchange);
        
        try {
            @SuppressWarnings("unchecked")
            Map<String, Object> request = gson.fromJson(requestBody, Map.class);
            
            // Prüfe, ob x und y im Request vorhanden sind
            if (!request.containsKey("x") || !request.containsKey("y")) {
                ApiRouter.sendResponse(exchange, 400, gson.toJson(Map.of("error", "x and y parameters are required")));
                return;
            }
            
            Object xObj = request.get("x");
            Object yObj = request.get("y");
            double x, y;
            
            if (xObj instanceof Number) {
                x = ((Number) xObj).doubleValue();
            } else if (xObj instanceof String) {
                x = Double.parseDouble((String) xObj);
            } else {
                ApiRouter.sendResponse(exchange, 400, gson.toJson(Map.of("error", "x must be a number")));
                return;
            }
            
            if (yObj instanceof Number) {
                y = ((Number) yObj).doubleValue();
            } else if (yObj instanceof String) {
                y = Double.parseDouble((String) yObj);
            } else {
                ApiRouter.sendResponse(exchange, 400, gson.toJson(Map.of("error", "y must be a number")));
                return;
            }
            
            // Validiere, dass x und y zwischen 0.0 und 1.0 liegen
            if (x < 0.0 || x > 1.0 || y < 0.0 || y > 1.0) {
                ApiRouter.sendResponse(exchange, 400, gson.toJson(Map.of("error", "x and y must be between 0.0 and 1.0")));
                return;
            }
            
            // Runde auf 3 Nachkommastellen
            x = Math.round(x * 1000.0) / 1000.0;
            y = Math.round(y * 1000.0) / 1000.0;
            
            // Rufe die Methode im HueModuleManager auf
            boolean success = hueModule.setColor(deviceId, x, y);
            if (success) {
                ApiRouter.sendResponse(exchange, 200, gson.toJson(Map.of("success", true)));
            } else {
                ApiRouter.sendResponse(exchange, 404, gson.toJson(Map.of("error", "Gerät nicht gefunden oder nicht unterstützt")));
            }
        } catch (Exception e) {
            logger.error("Fehler beim Setzen der Farbe", e);
            ApiRouter.sendResponse(exchange, 400, gson.toJson(Map.of("error", "Invalid request: " + e.getMessage())));
        }
    }
}

File: src/main/node/com/smarthome/backend/server/api/services/modules/hueModule.service.ts
import { Router } from "express";
import type { DatabaseManager } from "../../../db/database.js";
import type { EventStreamManager } from "../../../events/eventStreamManager.js";
import type { ActionManager } from "../../../actions/actionManager.js";
import { HueModuleManager } from "../../modules/hue/hueModuleManager.js";
import { logger } from "../../../../logger.js";

type Deps = {
  databaseManager: DatabaseManager;
  eventStreamManager: EventStreamManager;
  actionManager: ActionManager;
};

export function createHueModuleRouter(deps: Deps) {
  const router = Router();
  const hueModule = new HueModuleManager(deps.databaseManager, deps.eventStreamManager, deps.actionManager);

  const handleSetSensitivity = (req: { params: { deviceId: string }; body?: any }, res: any) => {
    try {
      logger.info("Setze Sensitivity fuer Hue Motion Sensor: {}", req.params.deviceId);
      const rawSensitivity = (req.body ?? {}).sensitivity;
      let sensitivity: number | null = null;
      if (typeof rawSensitivity === "number") {
        sensitivity = rawSensitivity;
      } else if (typeof rawSensitivity === "string" && rawSensitivity.trim().length > 0) {
        const parsed = Number(rawSensitivity);
        if (!Number.isNaN(parsed)) sensitivity = parsed;
      }
      if (sensitivity == null) {
        res.status(400).json({ error: "sensitivity parameter is required" });
        return;
      }
      const success = hueModule.setSensitivity(req.params.deviceId, sensitivity);
      res.status(success ? 200 : 404).json(success ? { success: true } : { error: "Gerät nicht gefunden" });
    } catch (error) {
      logger.error({ error }, "Fehler beim Setzen der Sensitivity");
      res.status(400).json({ error: "Invalid request" });
    }
  };

  const handleSetOn = (req: { params: { deviceId: string } }, res: any) => {
    try {
      logger.info("Schalte Hue Light ein: {}", req.params.deviceId);
      const success = hueModule.setOn(req.params.deviceId);
      res.status(success ? 200 : 404).json(success ? { success: true } : { error: "Gerät nicht gefunden oder nicht unterstützt" });
    } catch (error) {
      logger.error({ error }, "Fehler beim Einschalten des Hue Light");
      res.status(400).json({ error: "Invalid request" });
    }
  };

  const handleSetOff = (req: { params: { deviceId: string } }, res: any) => {
    try {
      logger.info("Schalte Hue Light aus: {}", req.params.deviceId);
      const success = hueModule.setOff(req.params.deviceId);
      res.status(success ? 200 : 404).json(success ? { success: true } : { error: "Gerät nicht gefunden oder nicht unterstützt" });
    } catch (error) {
      logger.error({ error }, "Fehler beim Ausschalten des Hue Light");
      res.status(400).json({ error: "Invalid request" });
    }
  };

  const handleSetBrightness = (req: { params: { deviceId: string }; body?: any }, res: any) => {
    try {
      logger.info("Setze Helligkeit fuer Hue Light: {}", req.params.deviceId);
      const rawBrightness = (req.body ?? {}).brightness;
      let brightness: number | null = null;
      if (typeof rawBrightness === "number") {
        brightness = rawBrightness;
      } else if (typeof rawBrightness === "string" && rawBrightness.trim().length > 0) {
        const parsed = Number(rawBrightness);
        if (!Number.isNaN(parsed)) brightness = parsed;
      }
      if (brightness == null) {
        res.status(400).json({ error: "brightness parameter is required" });
        return;
      }
      const success = hueModule.setBrightness(req.params.deviceId, brightness);
      res.status(success ? 200 : 404).json(success ? { success: true } : { error: "Gerät nicht gefunden oder nicht unterstützt" });
    } catch (error) {
      logger.error({ error }, "Fehler beim Setzen der Helligkeit");
      res.status(400).json({ error: "Invalid request" });
    }
  };

  const handleSetTemperature = (req: { params: { deviceId: string }; body?: any }, res: any) => {
    try {
      logger.info("Setze Farbtemperatur fuer Hue Light: {}", req.params.deviceId);
      const rawTemperature = (req.body ?? {}).temperature;
      let temperature: number | null = null;
      if (typeof rawTemperature === "number") {
        temperature = rawTemperature;
      } else if (typeof rawTemperature === "string" && rawTemperature.trim().length > 0) {
        const parsed = Number(rawTemperature);
        if (!Number.isNaN(parsed)) temperature = parsed;
      }
      if (temperature == null) {
        res.status(400).json({ error: "temperature parameter is required" });
        return;
      }
      const success = hueModule.setTemperature(req.params.deviceId, temperature);
      res.status(success ? 200 : 404).json(success ? { success: true } : { error: "Gerät nicht gefunden oder nicht unterstützt" });
    } catch (error) {
      logger.error({ error }, "Fehler beim Setzen der Farbtemperatur");
      res.status(400).json({ error: "Invalid request" });
    }
  };

  const handleSetColor = (req: { params: { deviceId: string }; body?: any }, res: any) => {
    try {
      logger.info("Setze Farbe fuer Hue Light: {}", req.params.deviceId);
      const rawX = (req.body ?? {}).x;
      const rawY = (req.body ?? {}).y;
      let x: number | null = null;
      let y: number | null = null;
      if (typeof rawX === "number") {
        x = rawX;
      } else if (typeof rawX === "string" && rawX.trim().length > 0) {
        const parsed = Number(rawX);
        if (!Number.isNaN(parsed)) x = parsed;
      }
      if (typeof rawY === "number") {
        y = rawY;
      } else if (typeof rawY === "string" && rawY.trim().length > 0) {
        const parsed = Number(rawY);
        if (!Number.isNaN(parsed)) y = parsed;
      }
      if (x == null || y == null) {
        res.status(400).json({ error: "x and y parameters are required" });
        return;
      }
      if (x < 0 || x > 1 || y < 0 || y > 1) {
        res.status(400).json({ error: "x and y must be between 0.0 and 1.0" });
        return;
      }
      const roundedX = Math.round(x * 1000) / 1000;
      const roundedY = Math.round(y * 1000) / 1000;
      const success = hueModule.setColor(req.params.deviceId, roundedX, roundedY);
      res.status(success ? 200 : 404).json(success ? { success: true } : { error: "Gerät nicht gefunden oder nicht unterstützt" });
    } catch (error) {
      logger.error({ error }, "Fehler beim Setzen der Farbe");
      res.status(400).json({ error: "Invalid request" });
    }
  };

  router.get("/bridges/discover", async (_req, res) => {
    try {
      const bridges = await hueModule.discoverBridges();
      res.status(200).json(bridges);
    } catch (error) {
      logger.error({ error }, "Fehler beim Discover von Hue-Bridges");
      res.status(500).json({ error: "Fehler beim Discover von Hue-Bridges" });
    }
  });

  router.post("/bridges/:bridgeId/pair", async (req, res) => {
    try {
      const result = await hueModule.pairBridge(req.params.bridgeId, req.body ?? {});
      res.status(result ? 200 : 404).json(result ? { success: true } : { error: "Bridge nicht gefunden" });
    } catch (error) {
      logger.error({ error }, "Fehler beim Pairing der Hue-Bridge");
      res.status(400).json({ error: "Invalid request" });
    }
  });

  router.get("/discover/devices/:bridgeId", async (req, res) => {
    try {
      const devices = await hueModule.discoverDevices(req.params.bridgeId);
      res.status(200).json(devices);
    } catch (error) {
      logger.error({ error }, "Fehler beim Discover von Hue-Geraeten");
      res.status(400).json({ error: "Invalid request" });
    }
  });

  router.post("/devices/:deviceId/setSensitivity", handleSetSensitivity);
  router.put("/devices/:deviceId/setSensitivity", handleSetSensitivity);

  router.post("/devices/:deviceId/setOn", handleSetOn);
  router.put("/devices/:deviceId/setOn", handleSetOn);

  router.post("/devices/:deviceId/setOff", handleSetOff);
  router.put("/devices/:deviceId/setOff", handleSetOff);

  router.post("/devices/:deviceId/setBrightness", handleSetBrightness);
  router.put("/devices/:deviceId/setBrightness", handleSetBrightness);

  router.post("/devices/:deviceId/setTemperature", handleSetTemperature);
  router.put("/devices/:deviceId/setTemperature", handleSetTemperature);

  router.post("/devices/:deviceId/setColor", handleSetColor);
  router.put("/devices/:deviceId/setColor", handleSetColor);

  return router;
}

